Fully implement Discord bot

Paldiu opened the pull request:
[Night Config]: https://github.com/TheElectronWill/night-config
[minecrell/plugin-yml]: https://github.com/Minecrell/plugin-yml
[TOML]: https://toml.io
[JSON]: https://json.org

[toml]: https://github.com/TheElectronWill/night-config/tree/master/toml
[json]: https://github.com/TheElectronWill/night-config/tree/master/json
[core]: https://github.com/TheElectronWill/night-config/tree/master/core

# Changelog:

- Created `Tyr` as a reference for future implementations targeting `SSH -> RCON service`.
- Created `Veritas`, a Discord bot that beams Discord messages to Minecraft, and vice versa. This project uses Discord4J, a reactive Rest/RS Java Wrapper for the Discord API based on the Reactive Streams specification.
- Added library dependencies for [Night Config] artifacts [toml], [json], and [core] to `Patchwork`.
- Add the appropriate Bukkit tags required by the [minecrell/plugin-yml] Gradle plugin to `Veritas` and `Tyr`.
- Created a `GenericConfiguration` class which uses [Night Config] to abstract away the specifics of working with language-specific semantics and wraps it nicely into a singular configuration class.
- Created the `ConfigType` enum to define configuration formats, parsers, and writers for [TOML] and [JSON], which is used by the `GenericConfiguration` class to determine the configuration language to target.
- Created `FileUtils` class containing useful file and directory creation methods
- Added JetBrains annotation `@ApiStatus.Internal` to both the `BukkitDelegate` class (also annotated with `@ApiStatus.NonExtendable`) and Completions annotation to specify that they should not be used outside of their current purpose within `Patchwork`.

# New Features

## Veritas
- Help command, which displays the currently supported commands
- TPS command, which displays the current Minecraft server TPS for the last 1 minute, 5 minutes, and 15 minutes 
- List command, which displays a list of all online players. Optionally, if the staff option is set to true, it will display only staff members.
- Backed by Discord4J, which provides the benefit of reactive non-blocking REST calls to minimize impact on the server tick loop (no need to force operating thread to wait for a response)
- JSON command registration (command backend implementation still required) for customizing which commands are loaded by the bot
- A select amount of features from the original TFM Discord Bot have been rewritten and implemented in this software as well

## Patchwork
- GenericConfiguration class can be used in conjunction with ConfigType enum to define either [TOML] or [JSON] configuration files. GenericConfiguration is an implementation of the Configuration interface which uses [Night Config] to interact with both file types without exposing any language-specific semantics.
- FileUtils class which is a utility class that provides various helpful methods for File interaction. Currently, FileUtils has `getOrCreateDirectory`, `getOrCreateFile`, `getOrCreateFileWithResource`, and `getExtension` methods. This will be expanded upon in the future.
- FileUtils `getOrCreateFile` and `getOrCreateFileWithResource` both return an `Optional<File>` to avoid null pointers.

Paldiu committed:
Fully implement Discord bot

ghost commented:
I think this pull request should be marked as a draft until the TOML configuration implementation is fixed and the merge conflicts are too.

Paldiu committed:
Merge remote-tracking branch 'origin/Implementations' into impl/discord-bot

Paldiu committed:
Fix plugin-yml plugin errors & switch to night-config

# Changes:
- Migrate from com.google.gson:gson and com.tomlj:tomlj over to com.electronwill.night-config:core, toml, and json
- Add the appropriate bukkit tags required by the minecrell/plugin-yml gradle plugin to Veritas and Tyr.

Paldiu committed:
Replace TOML wrapper with generic support

# Changes:
- Removed specific TOML wrapper in favor of GenericConfiguration.
- Added ConfigType enum to define configuration formats, parsers, and writers for TOML and JSON.
- Created FileUtils class containing useful file and directory creation methods
- Added @ApiStatus.Internal to both the BukkitDelegate class and Completions annotation to specify that they should not be used externally.
